algorithms/sort/insertion_sort.py

```diff
@@ -1,15 +1,24 @@
-def insertion_sort(arr):
+def insertion_sort(arr, simulation=False):
     """ Insertion Sort
         Complexity: O(n^2)
     """
+    if simulation:
+        print(arr)
+        
     for i in range(len(arr)):
         cursor = arr[i]
         pos = i
+        
+        temp = arr[:]
```

Comment:
I believe this makes the space complexity from O(1) to O(N).
I'm not sure about increasing the time complexity for visualization purpose.

Would there be a better way we could do this?

Reply:
Yeah, definitely there's a better way.
We can have a flag `swapped` to tell if any the position of any element has swapped.
for each `i` initialize `swapped = False`. Then if it enters the while loop and swaps, set `swapped = True`.
change the condition to
```python
if simulation and not swapped:
    print(arr)
```

Reply:
@lg970325 But I don't understand why are not printing the array when its not changed.
I think we should print status of array after each `iteration`, irrespective of whether its changed or not, because then it will show the accurate steps of what the algorithm is doing.

Reply:
@goswami-rahul What I thought when implementing simulation feature was to show how the array **changed** as sort algorithm processes. That's the reason why I didn't printed the array when its not changed. But printing the array after each iteration is more accurate as you said, so I'll change that part.

Reply:
I changed bubble, insertion and selection sort to print the array at every iteration. So in insertion sort for example, if the arr=[3,4,5,1,2], and you execute insertion_sort(arr,True), you will see:
iteration 0 : [3, 4, 5, 1, 2]
iteration 1 : [3, 4, 5, 1, 2]
iteration 2 : [3, 4, 5, 1, 2]
iteration 3 : [3, 4, 5, 1, 2]
iteration 4 : [1, 3, 4, 5, 2]
iteration 5 : [1, 2, 3, 4, 5]